waku/v2/api/filter.go

```diff
@@ -0,0 +1,225 @@
+package api
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/filter"
+	"github.com/waku-org/go-waku/waku/v2/protocol/subscription"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+)
+
+type FilterConfig struct {
+	MaxPeers int
+	Peers    []peer.ID
+}
+
+type Sub struct {
+	ContentFilter protocol.ContentFilter
+	DataCh        chan *protocol.Envelope
+	Config        FilterConfig
+	subs          subscription.SubscriptionSet
+	wf            *filter.WakuFilterLightNode
+	ctx           context.Context
+	cancel        context.CancelFunc
+	log           *zap.Logger
+}
+
+// Subscribe
+func Subscribe(ctx context.Context, wf *filter.WakuFilterLightNode, contentFilter protocol.ContentFilter, config FilterConfig) (*Sub, error) {
+	sub := new(Sub)
+	sub.wf = wf
+	sub.ctx, sub.cancel = context.WithCancel(ctx)
+	sub.subs = make(subscription.SubscriptionSet)
+	sub.DataCh = make(chan *protocol.Envelope)
+	sub.ContentFilter = contentFilter
+	sub.Config = config
+	sub.log = func() *zap.Logger {
+		log, _ := zap.NewDevelopment()
+		return log
+	}().Named("filterv2-api")
+
+	subs, err := sub.subscribe(contentFilter, sub.Config.MaxPeers)
+
+	if err == nil {
+		sub.multiplex(subs)
+		sub.log.Info("go sub.healthCheckLoop()")
+		go sub.healthCheckLoop()
+		return sub, nil
+	} else {
+		return nil, err
+	}
+}
+
+func (apiSub *Sub) Unsubscribe() {
+	apiSub.cancel()
+
+}
+
+func (apiSub *Sub) healthCheckLoop() {
+	// Health checks
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-apiSub.ctx.Done():
+			apiSub.log.Info("healthCheckLoop: Done()")
+			apiSub.cleanup()
+			return
+		case <-ticker.C:
+			apiSub.log.Info("healthCheckLoop: checkAliveness()")
+			topicCounts := apiSub.getTopicCounts()
+			apiSub.resubscribe(topicCounts)
+		}
+	}
+
+}
+
+func (apiSub *Sub) cleanup() {
+	apiSub.log.Info("ENTER cleanup()")
+	defer func() {
+		apiSub.log.Info("EXIT cleanup()")
+	}()
+
+	for _, s := range apiSub.subs {
+		apiSub.wf.UnsubscribeWithSubscription(apiSub.ctx, s)
+	}
+	close(apiSub.DataCh)
+
+}
+
+// Returns active sub counts for each pubsub topic
+func (apiSub *Sub) getTopicCounts() map[string]int {
+	apiSub.log.Info("ENTER getTopicCounts()")
+	defer func() {
+		apiSub.log.Info("EXIT getTopicCounts()")
+	}()
+
+	// Buffered chan for sub aliveness results
+	type CheckResult struct {
+		sub   *subscription.SubscriptionDetails
+		alive bool
+	}
+	checkResults := make(chan CheckResult, len(apiSub.subs))
+
+	// Run pings asynchronously
+	for _, s := range apiSub.subs {
+		go func() {
+			ctx, _ := context.WithTimeout(apiSub.ctx, 5*time.Second)
+			err := apiSub.wf.IsSubscriptionAlive(ctx, s)
```

Comment:
One problem i see with this approach is that we may end up pinging each peer multiple times with Subscriber_Ping for same shard/pubsubTopic. Consider this scenario e.g: 

1. User joins a community (i.e a new shard or pubsubTopic PT1) and gets access to maybe public channels..let's say 5 contentTopics CT1 to CT5. We will subscribe using Filter Sub1 (PT1 , CT1...CT5)
2. After sometime, user is given access to some other channels which are on different topics i.e CT6, CT7. So another Filter subscription would happen for Sub2(PT1, CT6, CT7) 
3. During health-check 2 Filter Pings would go for both these Subs which is un-necessary. 

Filter-Ping should go only once towards a peer for all subscriptions combined (whether they are same or different pubsubTopics).

Reply:
Since the underlying Subscriptions have the peerId, maybe we can identify unique peers and send pings to them. If any ping fails, then the relevant subscriptions needs to be resubscribed.

Reply:
Agree, this should be optimized.

Reply:
Will take this optimization as part of a separate PR.

Reply:
Tracking as part of https://github.com/waku-org/go-waku/issues/1099